### PR TITLE
Add an option not to resolve source when symlinking.

### PR DESCRIPTION
**Pull request by David-5-1**
When the links are willingly made, the option allow forcing copyfile to create a symlink to the source without resolving its path.

In our setup we have a link for each major-minor python version which points to a specific build.
With current copyfile, this results in links to the sepcific build in created virtualenvs so we cannot remove any build unless when ensure all virtualenvs are no longer using it.
This MR adds a way to create symlinks to those symlinks, while keeping the default of resolving the symlinks (which was done to address issues with certain setups).

For now, my goal is mainly to get feedback on the proposed change, so I did not attempt to add documentation or edit changelog.

### CODE REVIEW

**Comment by David-5-1** [comment]
In our setup we have a link for each major-minor python version which points to a specific build.
With current copyfile, this results in links to the sepcific build in created virtualenvs so we cannot remove any build unless when ensure all virtualenvs are no longer using it.
This MR adds a way to create symlinks to those symlinks, while keeping the default of resolving the symlinks (which was done to address issues with certain setups).

For now, my goal is mainly to get feedback on the proposed change, so I did not attempt to add documentation or edit changelog.

**Comment by David-5-1** [comment]
My changes to tests were breaking on Windows.
As `copyfile` never creates symlinks on Windows, I disabled the added parts of the tests on Windows…

File: tests/test_virtualenv.py [hunk 1]
```diff
@@ -370,9 +371,20 @@ def test_copyfile_from_symlink(tmp_path):
...
-    virtualenv.copyfile(str(src_dir / "link2"), str(copy_path))
+
+    # Check with symlink_resolve True then False
+    virtualenv.copyfile(str(src_dir / "link2"), str(copy_path), symlink_resolve=True)
+    with open(str(copy_path), "r") as f:
+        assert f.read() == "contents"
+    if not virtualenv.IS_WIN:
+        assert os.readlink(str(copy_path)) == str(src_dir / "file")
+
+    os.remove(str(copy_path))
```
Comment (gaborbernat) [inline tests/test_virtualenv.py]: I think this test should be separate from this 👍 and we should leave untouched this test 👍

Comment (David-5-1) [reply tests/test_virtualenv.py]: With theses changes, we get 4 tests for copyfile.
To be sure I understand, do you suggest having two test functions, one using `copyfile` with `symlink_resolve` set to `True` and one where it is `False` ?

**Comment by David-5-1** [comment]
As you guessed, we are using the symlinks to builds to upgrade versions. We are providing a few different “major” releases (2.7, 3.6, 3.7) and upgrading them as new minor releases come.
From our setup, the new python minor version will be only used when we restart uWSGI service so it is as if the virtualenv was liked to a system python install which would be upgraded.
Major python upgrades would likely however break things when there are incompatible API changes (I can think of `async` and `await` that were added as keywords recently).

File: tests/test_virtualenv.py (hunk 1 above)
Comment (gaborbernat) [reply tests/test_virtualenv.py]: I would like to have these four test case as separate test functions, and please add a changelog that explains the need and addition.

**Comment by gaborbernat** [comment]
I see your use case, in practice I've seen it applied more along the lines that all server packages are rebuilt on minor interpreter release. Feels a bit safer to do so.

**Comment by asottile** [comment]
in my experience, patch upgrades usually break virtualenvs (for example, try making a 3.6.2 virtualenv against symlinks and then upgrading to 3.6.3) -- I'm surprised you've had success with this approach

**Comment by gaborbernat** [comment]
Closing due to no action on the PR for a long time, if anyone has the bandwidth to pick this up, let me know, and we can have a talk what would be a good solution here.

**Pull request closed without merging**